### PR TITLE
2x faster (rms) norm cuda kernels (3.7% e2e improvement)

### PR DESCRIPTION
**Pull request by li-plus**
This PR optimized rms_norm and norm cuda kernels for hidden size over 1024, achieving a 2x speedup on kernel and a 3.7% end-to-end generation speedup.

Tested on a V100 GPU with a LLaMA-7B q4_0 model with 4096 hidden size:
```sh
./main -m ./models/7B/ggml-model-q4_0.gguf -p "Building a website can be done in 10 simple steps:" -n 512 -ngl 32 -t 1
```

Before this PR:
```
llama_print_timings:        load time =  1265.11 ms
llama_print_timings:      sample time =   298.76 ms /   512 runs   (    0.58 ms per token,  1713.74 tokens per second)
llama_print_timings: prompt eval time =   290.99 ms /   271 tokens (    1.07 ms per token,   931.30 tokens per second)
llama_print_timings:        eval time =  4628.23 ms /   510 runs   (    9.07 ms per token,   110.19 tokens per second)
llama_print_timings:       total time =  5443.79 ms
```
Single rms_norm kernel time: 8.9us (reported by nsys)

After this PR:
```
llama_print_timings:        load time =  1283.14 ms
llama_print_timings:      sample time =   306.41 ms /   512 runs   (    0.60 ms per token,  1670.99 tokens per second)
llama_print_timings: prompt eval time =   287.84 ms /   271 tokens (    1.06 ms per token,   941.49 tokens per second)
llama_print_timings:        eval time =  4452.48 ms /   510 runs   (    8.73 ms per token,   114.54 tokens per second)
llama_print_timings:       total time =  5268.04 ms
```
Single rms_norm kernel time: 4.5us (reported by nsys)

Kernel speedup: 8.9us/4.5us = 1.98

Per token latency reduced by: 1-(8.73ms/9.07ms) = 3.7%

Similar speedup rate is achieved for norm op tested with a ChatGLM-6B q4_0 model using [chatglm.cpp](https://github.com/li-plus/chatglm.cpp): 8.1 ms/token before this PR v.s. 7.8 ms/token with this PR (3.7% improvement)

### CODE REVIEW

File: ggml-cuda.cu [hunk 1]
```diff
@@ -464,58 +464,91 @@ static __global__ void silu_f32(const float * x, float * dst, const int k) {
...
+static __device__ __forceinline__ float2 warp_reduce_sum(float2 a) {
+#pragma unroll
+    for (int mask = 16; mask > 0; mask >>= 1) {
+        a.x += __shfl_xor_sync(0xffffffff, a.x, mask, 32);
+        a.y += __shfl_xor_sync(0xffffffff, a.y, mask, 32);
+    }
+    return a;
+}
+
+template <int BLOCK_SIZE>
```
Comment (JohannesGaessler) [inline ggml-cuda.cu:476]: To keep the style consistent:
```suggestion
template <int block_size>
```

File: ggml-cuda.cu [hunk 2]
```diff
@@ -464,58 +464,91 @@ static __global__ void silu_f32(const float * x, float * dst, const int k) {
...
 
+static __device__ __forceinline__ float warp_reduce_sum(float x) {
+#pragma unroll
+    for (int mask = 16; mask > 0; mask >>= 1) {
+        x += __shfl_xor_sync(0xffffffff, x, mask, 32);
+    }
+    return x;
+}
+
+template <int BLOCK_SIZE>
```
Comment (JohannesGaessler) [inline ggml-cuda.cu:522]: Same as above.

File: ggml-cuda.cu [hunk 3]
```diff
@@ -4203,14 +4236,24 @@ static void silu_f32_cuda(const float * x, float * dst, const int k, cudaStream_
 
 static void norm_f32_cuda(const float * x, float * dst, const int ncols, const int nrows, cudaStream_t stream) {
     GGML_ASSERT(ncols % WARP_SIZE == 0);
-    const dim3 block_dims(WARP_SIZE, 1, 1);
-    norm_f32<<<nrows, block_dims, 0, stream>>>(x, dst, ncols);
+    if (ncols < 1024) {
```
Comment (JohannesGaessler) [inline ggml-cuda.cu:4239]: I think the code will not produce correct results for tensor sizes not divisible by 1024. So I think the check should instead be:

```suggestion
    if (ncols % 1024 != 0) {
```

File: ggml-cuda.cu [hunk 4]
```diff
@@ -4203,14 +4236,24 @@ static void silu_f32_cuda(const float * x, float * dst, const int k, cudaStream_
...
+        const dim3 block_dims(1024, 1, 1);
+        norm_f32<1024><<<nrows, block_dims, 0, stream>>>(x, dst, ncols);
+    }
 }
 
 static void rms_norm_f32_cuda(const float * x, float * dst, const int ncols, const int nrows, const float eps, cudaStream_t stream) {
     GGML_ASSERT(ncols % WARP_SIZE == 0);
-    const dim3 block_dims(WARP_SIZE, 1, 1);
-    rms_norm_f32<<<nrows, block_dims, 0, stream>>>(x, dst, ncols, eps);
+    if (ncols < 1024) {
```
Comment (JohannesGaessler) [inline ggml-cuda.cu:4250]: Same as above:

```suggestion
    if (ncols % 1024 != 0) {
```

File: ggml-cuda.cu (hunk 3 above)
Comment (li-plus) [reply ggml-cuda.cu:4239]: It produces correct results for any hidden size. I tested it with hidden size 1234 and it produces the same results as PyTorch. I think we should even remove this check `GGML_ASSERT(ncols % WARP_SIZE == 0);`.

File: ggml-cuda.cu (hunk 1 above)
Comment (li-plus) [reply ggml-cuda.cu]: Fixed. Made it lowercase.

File: ggml-cuda.cu (hunk 2 above)
Comment (li-plus) [reply ggml-cuda.cu]: Fixed as well

File: ggml-cuda.cu (hunk 3 above)
Comment (JohannesGaessler) [reply ggml-cuda.cu:4239]: Yes, I think you're right. The loops only go up to `ncols` so this should make out-of-bounds memory accesses impossible.

**Review by JohannesGaessler — APPROVED** [review]
I checked the performance:

| Model                | GPU                     | test   |   t/s master |   t/s PR |   Speedup |
|----------------------|-------------------------|--------|--------------|----------|-----------|
| LLaMA 7B mostly Q4_0 | NVIDIA GeForce RTX 3090 | pp512  |      2175.11 |  2206.13 |      1.01 |
| LLaMA 7B mostly Q4_0 | NVIDIA GeForce RTX 3090 | tg128  |       130.45 |   134.08 |      1.03 |
| LLaMA 7B mostly Q4_0 | AMD Radeon RX 6800 | pp512  |      1054.46 |  1057.74 |      1.00 |
| LLaMA 7B mostly Q4_0 | AMD Radeon RX 6800 | tg128  |        64.57 |    71.86 |      1.11 |
| LLaMA 7B mostly Q4_0 | Tesla P40          | pp512  |       871.55 |   874.11 |      1.00 |
| LLaMA 7B mostly Q4_0 | Tesla P40          | tg128  |        51.93 |    54.76 |      1.05 |

I'll leave this PR open for a few more hours in case someone wants to object but I think this can be merged. You could maybe move the `warp_reduce_sum` helper functions upwards and use them across all kernels but this is a quick and simple change that I could just do myself.

**Pull request merged**